test/Stackctl/AWS/CloudFormationSpec.hs

```diff
@@ -2,27 +2,95 @@ module Stackctl.AWS.CloudFormationSpec
 import Stackctl.AWS.CloudFormation
-import Test.Hspec
 
 spec :: Spec
 spec = do
   describe "readParameter" $ do
-    it "refuses empty key" $ do
+    it "refuses empty key" $ example $ do
       readParameter "=Value"
         `shouldSatisfy` either ("empty KEY" `isSuffixOf`) (const False)
 
-    it "refuses empty value" $ do
+    it "refuses empty value" $ example $ do
       readParameter "Key"
         `shouldSatisfy` either ("empty VALUE" `isSuffixOf`) (const False)
 
-    it "refuses empty value (with =)" $ do
+    it "refuses empty value (with =)" $ example $ do
       readParameter "Key="
         `shouldSatisfy` either ("empty VALUE" `isSuffixOf`) (const False)
 
-    it "creates a parameter when valid" $ do
+    it "creates a parameter when valid" $ example $ do
       readParameter "Key=Value=More"
         `shouldBe` Right (makeParameter "Key" $ Just "Value=More")
+
+  describe "awsCloudFormationDeleteAllChangeSets" $ do
+    it "deletes all listed changesets" $ example $ runTestAppT $ do
+      let
+        stackName :: Text
+        stackName = "some-stack"
+
+        cs1 :: Text
+        cs1 = "some-changeset-1"
+
+        cs2 :: Text
+        cs2 = "some-changeset-2"
+
+        cs3 :: Text
+        cs3 = "some-changeset-3"
+
+        isListChangeSetsPage :: Maybe Text -> ListChangeSets -> Bool
+        isListChangeSetsPage p req =
+          and
+            [ req ^. listChangeSets_stackName == stackName
+            , req ^. listChangeSets_nextToken == p
+            ]
+
+        isDeleteChangeSet :: Text -> DeleteChangeSet -> Bool
+        isDeleteChangeSet cs req = req ^. deleteChangeSet_changeSetName == cs
+
+        summary1 = newChangeSetSummary & changeSetSummary_changeSetId ?~ cs1
+        summary2 = newChangeSetSummary & changeSetSummary_changeSetId ?~ cs2
+        summary3 = newChangeSetSummary & changeSetSummary_changeSetId ?~ cs3
+
+        matchers =
+          [ SendMatcher (isListChangeSetsPage Nothing)
+              $ Right
+              $ newListChangeSetsResponse 200
+              & (listChangeSetsResponse_summaries ?~ [summary1, summary2])
+              & (listChangeSetsResponse_nextToken ?~ "p2")
+          , SendMatcher (isListChangeSetsPage $ Just "p2")
+              $ Right
+              $ newListChangeSetsResponse 200
+              & (listChangeSetsResponse_summaries ?~ [summary3])
+          , SendMatcher (isDeleteChangeSet cs1)
+              $ Right
+              $ newDeleteChangeSetResponse 200
+          , SendMatcher (isDeleteChangeSet cs2)
+              $ Right
+              $ newDeleteChangeSetResponse 200
+          , SendMatcher (isDeleteChangeSet cs3)
+              $ Right
+              $ newDeleteChangeSetResponse 200
+          ]
+
+      withMatchers matchers $ do
+        awsCloudFormationDeleteAllChangeSets $ StackName stackName
+
+        messages <-
+          map (loggedMessageText &&& loggedMessageMeta)
+            <$> getLoggedMessagesUnsafe
+
+        messages
```

Comment:
would it be appropriate to have a `LogMessage` type that these could match against, s.t. we're matching sum types here and not strings?

Reply:
I'm not sure what that would look like, can you say more?

Reply:
yeah so I suppose that if you did what I am suggesting the way I was thinking of it, you'd probably want to change "Blammo" somehow, but the rough idea is that there would be a type like:

```haskell
newtype ChangeSetId = ChangeSetId Text

data LogMessage 
  = DeleteAllChangesets
  | EnqueueingChangeSetDelete ChangeSetId
  | ...
  | UnknownLogMessage Text KeyMap
```

and then the actual assertion would be like:
```
 `shouldBe` [ DeletingAllChangesets
                     ,  EnqueueingChangeSetDelete cs1
                     ,  EnqueueingChangeSetDelete cs2
                     ,  EnqueueingChangeSetDelete cs3
                     ]
```

So it woudl be some kind of notion of typed log output. 

Just ideas, no need to actually do it here.

Reply:
> you'd probably want to change "Blammo" somehow

Yeah, this is what I can't really picture. Does every `log{Level}` function take some enum type? How does the library know about it when it seems app-specific. Happy to explore the idea, but I think you'll need to describe what that side of it would look like more.

Reply:
Yea so I didn't think about the log levels; I suppose there are two choices really, one type per log level, or all log levels share a type. I dunno which one would be preferable.

On the library side, Logger could take a type parameter, representing the type of the thing that is being logged. And likely functions would need some typeclass annotations e.g. functions would look like 

```
newTypedTestLogger :: IsLogMessage t => LogSettings -> TypedLogger t
```

And likely there'd need to be a typed version of `LoggedMessage`:

```
data TypedLoggedMessage t = TypedLoggedMessage LoggedMessage t
```

Do I make sense yet?

Reply:
If I don't, I can try forking blammo and seeing how it goes, and then we can discuss in PR-land.

Reply:
I think you're making sense. If you want to advance the idea, then moving this to a Blammo PR would be a good next step.